### PR TITLE
Fixes #13010 by updating boto3 methods with profile_name

### LINKED ISSUE #13010: AWS boto3 - profile support
**Issue Type:**

Bug Report

**Ansible Version**
ansible 2.0.0 (detached HEAD de54f26376) last updated 2015/11/03 08:46:06 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD e318be30ff) last updated 2015/11/03 08:45:42 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD 80f198a82f) last updated 2015/11/03 08:45:42 (GMT +1000)

Running Ansible from source - v2.0.0-0.4.beta2

**Environment:**
OSX 10.10.5
Running Ansible from source - v2.0.0-0.4.beta2
Running boto (2.38.0) and boto3 (1.1.4)
Python 2.7.10

**Summary:**

When utilising boto3_conn from module_utils/ec2.py and specifying a boto profile_name, the connection fails due to profile_name not being a valid boto3.session.Session().client() parameter.

_TypeError: client() got an unexpected keyword argument 'profile_name'_

**Steps To Reproduce:**

Creating a new module using boto3 for listing Route53 entries, host zones etc and allowing the user to provide a boto profile - same behaviour as in boto (version 2).

Connection setup in new module:

```
try:
    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
    route53 = boto3_conn(module, conn_type='client', resource='route53', **aws_connect_kwargs)
except boto.exception.NoAuthHandlerFound, e:
    module.fail_json(msg="Can't authorize connection - "+str(e))
```

boto3_conn section in question:

```
elif conn_type == 'client':
    client = boto3.session.Session().client(resource, region_name=region, endpoint_url=endpoint, **params)
return client
```

Setup a play using the new module and specify a profile: _name_

**Expected Results:**

```
TASK [route53_entries : Testing custom module] *********************************
 ok: [127.0.0.1] => {"HostedZones": [{ dictionary of returned hosted zones from AWS }}
```

**Actual Results:**

```
TASK [route53_entries : Testing custom module] *********************************
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/Etherdaemon/.ansible/tmp/ansible-tmp-1446523327.09-207574321397669/route53_facts", line 2208, in <module>
main()
  File "/Users/Etherdaemon/.ansible/tmp/ansible-tmp-1446523327.09-207574321397669/route53_facts", line 67, in main
    route53 = boto3_conn(module, conn_type='client', resource='route53', **aws_connect_kwargs)
  File "/Users/Etherdaemon/.ansible/tmp/ansible-tmp-1446523327.09-207574321397669/route53_facts", line 2033, in boto3_conn
    client = boto3.session.Session().client(resource, region_name=region, endpoint_url=endpoint, profile_name=profile_name, **params)
TypeError: client() got an unexpected keyword argument 'profile_name'

fatal: [127.0.0.1]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"profile": "personal", "type": "A"}, "module_name": "route53_facts"}, "parsed": false}
```

**Proposed fix:**

My initial thoughts are to do the following - I've tested this for just my own scenario so obviously a lot more thought has to go in for everyone's use cases.

Update module_utils/ec2.py with the following:
- method: get_aws_connection_info
  - Remove profile_name from boto_params dictionary for boto3 requests
  - move existing return region, ec2_url, boto_params to be directly underneath the boto/boto3 else statement (to cater for boto 2 connections)
  - add a new return statement under the boto3 if statement to include the new profile_name argument separately - return region, ec2_url, profile_name, boto_params
- method: boto3_conn
  - add a new profile_name=None input parameter
  - update the boto3.session statement to create the session with the profile and the client with the rest of the arguments.

This should then allow users to call the boto3_conn like so...

```
try:
    region, ec2_url, profile_name, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
    route53 = boto3_conn(module, conn_type='client', resource='route53', profile_name=profile_name, **aws_connect_kwargs)
except boto.exception.NoAuthHandlerFound, e:
    module.fail_json(msg="Can't authorize connection - "+str(e))
```

### PR DESCRIPTION
**Pull request by Etherdaemon**
Fixes #13010 by updating boto3 methods to have a separate profile_name argument as specified in the proposed fix section of the issue

### CODE REVIEW

File: lib/ansible/module_utils/ec2.py [hunk 1]
```diff
@@ -150,10 +150,8 @@ def get_aws_connection_info(module, boto3=False):
         if validate_certs:
             boto_params['verify'] = validate_certs
 
-        if profile_name:
-            boto_params['profile_name'] = profile_name
-
-
+        return region, ec2_url, profile_name, boto_params
```
Comment (willthames) [inline lib/ansible/module_utils/ec2.py]: This appears to change the specification of get_aws_connection_info to return 4 values only when using boto3. 

Wouldn't you need to update every single module that calls `get_aws_connection_info`?

File: lib/ansible/module_utils/ec2.py [hunk 2]
```diff
@@ -168,8 +166,8 @@ def get_aws_connection_info(module, boto3=False):
 
         if validate_certs and HAS_LOOSE_VERSION and LooseVersion(boto.Version) >= LooseVersion("2.6.0"):
             boto_params['validate_certs'] = validate_certs
-
-    return region, ec2_url, boto_params
+
```
Comment (willthames) [inline lib/ansible/module_utils/ec2.py]: Empty lines should be empty

**Comment by willthames** [comment]
Sorry, I commented without reading #13010. I'll comment there instead.

**Comment by jimi-c** [comment]
@Etherdaemon could you please squash these down to a single commit, and clean up the merges that are in this branch? Thanks!

**Comment by Etherdaemon** [comment]
No worries - I'll try and fix it up later today.

**Comment by Etherdaemon** [comment]
Hmm - massive fail. Might be easier to close and reopen a new PR instead with just the 1 change.

**Comment by willthames** [comment]
@Etherdaemon for future reference

```
git fetch ansibleremote devel
git rebase -i devel
```

change all but the first of the `pick`s to `fixup` (or `f`) (or `squash`/`s`) if you want to incorporate elements of that commit message in your overall commit). 

Then `git push -f yourremote yourbranch`

**Comment by Etherdaemon** [comment]
Thanks @willthames. I had screwed it by doing merges and pushes etc. And I was further screwed by my commits being between multiple other peoples commits - so although I think the above was fine, I went ahead and closed this one,created a nice clean one and created a new pull request - https://github.com/ansible/ansible/pull/13053 so that there was no risk of squishing/crapping over other people's commits.

**Pull request closed without merging**